thinc/loss.py

```diff
@@ -58,30 +58,34 @@ def __init__(
         else:
             self._name_to_i = {}
 
-    def convert_truths(self, truths, guesses: Floats2d) -> Tuple[Floats2d, Floats2d]:
+    def convert_truths(
+        self, truths: IntsOrFloatsOrStrs, guesses: Floats2d
+    ) -> Tuple[Floats2d, Floats2d]:
         xp = get_array_module(guesses)
         missing = []
         negatives_mask = None
         if self.names:
             negatives_mask = xp.ones((len(truths), len(self.names)), dtype="f")
         missing_value = self.missing_value
-        # Convert list of ints or list of strings
         if isinstance(truths, list):
-            truths = list(truths)
             if len(truths):
+                # Collect missing values for List[int]
                 if isinstance(truths[0], int):
                     for i, value in enumerate(truths):
                         if value == missing_value:
                             missing.append(i)
                 else:
                     if self.names is None:
-                        msg = (
+                        raise ValueError(
                             "Cannot calculate loss from list of strings without names. "
                             "You can pass the names as a keyword argument when you "
                             "create the loss object, "
                             "e.g. CategoricalCrossentropy(names=['dog', 'cat'])"
                         )
-                        raise ValueError(msg)
+                    # Make sure "truths" is List[str]
+                    assert isinstance(truths[0], str)
```

Comment:
This is not part of the user input validation. In `convert_truths` there is not much input validation because now the typing is "fixed" in the sense that the `IntsOrFloatsOrStrs` is the type of `truths` so if the user provides the right type then this works. The `assert` is there for clarity that if it wasn't `List[int]` then at this point of the code it must be a `List[str]`.

Reply:
But you'll only capture it if you are using MyPy? I think many downstream users don't use type checking, so their code will happily run until they get an assertion and it's not really clear from that what they did wrong.